File: backend/app/models/niche.rb
# Represents a market niche identified from trend signals.
# Niches are evaluated for their commercial potential based on demand/supply metrics
# from sources like AO3 (fanfiction) and Etsy (marketplace).
#
# @attr [Integer] id Primary key
# @attr [Integer] trend_signal_id Foreign key to TrendSignal
# @attr [String] name Name of the niche
# @attr [Text] description Detailed description of the niche
# @attr [String] community_type Type of community (e.g., 'fandom', 'subculture')
# @attr [Float] demand_score Calculated demand metric
# @attr [Float] supply_score Calculated supply metric
# @attr [Float] demand_supply_ratio Ratio of demand to supply (higher is better)
# @attr [Integer] ao3_works_count Number of fanfiction works on AO3
# @attr [Float] ao3_growth_rate Growth rate of AO3 works
# @attr [Integer] etsy_listing_count Number of similar listings on Etsy
# @attr [String] status Current pipeline status
# @attr [DateTime] discovered_at When the niche was first identified
# @attr [DateTime] created_at Record creation timestamp
# @attr [DateTime] updated_at Record update timestamp
class Niche < ApplicationRecord
  belongs_to :trend_signal
  has_many :cultural_tokens, dependent: :destroy

  validates :name, presence: true
  validates :status, presence: true

  # Pipeline status tracking niche through discovery to production
  # @note Default status is :discovered
  enum :status, {
    discovered: "discovered",
    evaluating: "evaluating",
    mining_tokens: "mining_tokens",
    generating_designs: "generating_designs",
    active: "active",
    declining: "declining",
    archived: "archived"
  }, default: :discovered

  # @!scope class
  # Returns niches sorted by demand/supply ratio in descending order
  # @return [ActiveRecord::Relation<Niche>] Niches ordered by commercial potential
  scope :by_demand_ratio, -> { order(demand_supply_ratio: :desc) }

  # @!scope class
  # Returns niches currently in the active pipeline
  # @return [ActiveRecord::Relation<Niche>] Niches that are not declining or archived
  scope :active_pipeline, -> { where(status: [:discovered, :evaluating, :mining_tokens, :generating_designs, :active]) }
end

File: frontend/src/types/niche.ts
export const NICHE_STATUSES = [
  'discovered',
  'evaluating',
  'mining_tokens',
  'generating_designs',
  'active',
  'declining',
  'archived',
] as const

export type NicheStatus = (typeof NICHE_STATUSES)[number]

export const STATUS_LABELS: Record<NicheStatus, string> = {
  discovered: 'Discovered',
  evaluating: 'Evaluating',
  mining_tokens: 'Mining Tokens',
  generating_designs: 'Generating Designs',
  active: 'Active',
  declining: 'Declining',
  archived: 'Archived',
}

export const STATUS_COLORS: Record<NicheStatus, string> = {
  discovered: 'border-l-blue-400',
  evaluating: 'border-l-amber-400',
  mining_tokens: 'border-l-purple-400',
  generating_designs: 'border-l-pink-400',
  active: 'border-l-green-400',
  declining: 'border-l-orange-400',
  archived: 'border-l-gray-400',
}

export const STATUS_BG_COLORS: Record<NicheStatus, string> = {
  discovered: 'bg-blue-400/10',
  evaluating: 'bg-amber-400/10',
  mining_tokens: 'bg-purple-400/10',
  generating_designs: 'bg-pink-400/10',
  active: 'bg-green-400/10',
  declining: 'bg-orange-400/10',
  archived: 'bg-gray-400/10',
}

export type CommunityType = 'fandom' | 'activist' | 'meme' | 'professional'

export const COMMUNITY_COLORS: Record<string, string> = {
  fandom: 'bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-200',
  activist: 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-200',
  meme: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-200',
  professional: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-200',
}

export interface Niche {
  id: number
  trend_signal_id: number
  name: string
  description: string
  community_type: string
  demand_score: number
  supply_score: number
  demand_supply_ratio: number
  ao3_works_count: number
  ao3_growth_rate: number
  etsy_listing_count: number
  status: NicheStatus
  discovered_at: string
  created_at: string
  updated_at: string
}

export interface NicheScorecard {
  niche: Niche
  demand_score: number
  supply_score: number
  demand_supply_ratio: number
  ao3_metrics: {
    works_count: number
    growth_rate: number
  }
  etsy_listing_count: number
}

export interface CulturalToken {
  id: number
  niche_id: number
  name: string
  token_type: string
  description: string
  relevance_score: number
  status: string
  created_at: string
  updated_at: string
}

export interface Design {
  id: number
  cultural_token_id: number
  name: string
  description: string
  image_url: string
  status: string
  created_at: string
  updated_at: string
}
